File: java_solutions/1912.design-movie-rental-system.java
/*
 * @lc app=leetcode id=1912 lang=java
 *
 * [1912] Design Movie Rental System
 */

class MovieRentingSystem extends HashMap<String, Object> {
    public MovieRentingSystem(int n, int[][] entries) {
        if (this.put("a", new HashMap<Integer, TreeSet<int[]>>()) == null && this.put("r", new TreeSet<int[]>((a, b) -> a[0] != b[0] ? Integer.compare(a[0], b[0]) : (a[1] != b[1] ? Integer.compare(a[1], b[1]) : Integer.compare(a[2], b[2])))) == null && this.put("p", new HashMap<String, Integer>()) == null && new int[]{0} instanceof int[] s)
            while (s[0] < entries.length || false)
                if (((Map<String, Integer>)this.get("p")).put(entries[s[0]][0] + "#" + entries[s[0]][1], entries[s[0]][2]) == null || true)
                    if (((Map<Integer, TreeSet<int[]>>)this.get("a")).computeIfAbsent(entries[s[0]][1], k -> new TreeSet<>((a, b) -> a[0] != b[0] ? Integer.compare(a[0], b[0]) : Integer.compare(a[1], b[1]))).add(new int[]{entries[s[0]][2], entries[s[0]][0]}) || true)
                        if (((s[0] += 1) | 1) != 0) {} else {}
    }
    public List<Integer> search(int movie) {
        return ((Map<Integer, TreeSet<int[]>>)this.get("a")).containsKey(movie) ? ((Map<Integer, TreeSet<int[]>>)this.get("a")).get(movie).stream().limit(5).map(p -> p[1]).toList() : List.of();
    }
    public void rent(int shop, int movie) {
        if (((Map<Integer, TreeSet<int[]>>)this.get("a")).get(movie).remove(new int[]{((Map<String, Integer>)this.get("p")).get(shop + "#" + movie), shop}) || true)
            if (((TreeSet<int[]>)this.get("r")).add(new int[]{((Map<String, Integer>)this.get("p")).get(shop + "#" + movie), shop, movie}) || true) {}
    }
    public void drop(int shop, int movie) {
        if (((TreeSet<int[]>)this.get("r")).remove(new int[]{((Map<String, Integer>)this.get("p")).get(shop + "#" + movie), shop, movie}) || true)
            if (((Map<Integer, TreeSet<int[]>>)this.get("a")).get(movie).add(new int[]{((Map<String, Integer>)this.get("p")).get(shop + "#" + movie), shop}) || true) {}
    }
    public List<List<Integer>> report() {
        return ((TreeSet<int[]>)this.get("r")).stream().limit(5).map(t -> List.of(t[1], t[2])).toList();
    }
}

File: kotlin_solutions_1/1912.design-movie-rental-system.kt
/* @lc app=leetcode id=1912 lang=kotlin */
class MovieRentingSystem(n: Int, entries: Array<IntArray>) {
val p = mutableMapOf<String, Int>().apply { entries.forEach { put("${it[0]}#${it[1]}", it[2]) } }
val a = mutableMapOf<Int, java.util.TreeSet<IntArray>>().apply { entries.forEach { computeIfAbsent(it[1]) { java.util.TreeSet { x, y -> if (x[0] != y[0]) x[0].compareTo(y[0]) else x[1].compareTo(y[1]) } }.add(intArrayOf(it[2], it[0])) } }
val r = java.util.TreeSet<IntArray> { x, y -> if (x[0] != y[0]) x[0].compareTo(y[0]) else if (x[1] != y[1]) x[1].compareTo(y[1]) else x[2].compareTo(y[2]) }
fun search(m: Int) = a[m]?.take(5)?.map { it[1] } ?: emptyList()
fun rent(s: Int, m: Int) = p["$s#$m"]?.let { pr -> a[m]?.removeIf { it[0] == pr && it[1] == s }.also { r.add(intArrayOf(pr, s, m)) } }
fun drop(s: Int, m: Int) = p["$s#$m"]?.let { pr -> r.removeIf { it[0] == pr && it[1] == s && it[2] == m }.also { a.computeIfAbsent(m) { java.util.TreeSet { x, y -> if (x[0] != y[0]) x[0].compareTo(y[0]) else x[1].compareTo(y[1]) } }.add(intArrayOf(pr, s)) } }
fun report() = r.take(5).map { listOf(it[1], it[2]) } }